crates/pop-fork/src/cache.rs

```diff
@@ -0,0 +1,564 @@
+	/// Creates the parent directory if it doesn't exist.
+	pub async fn open(path: &Path) -> Result<Self, CacheError> {
+		// Ensure parent directory exists
+		if let Some(parent) = path.parent() {
+			std::fs::create_dir_all(parent)?;
+		}
+
+		let url = format!("sqlite:{}?mode=rwc", path.display());
+		let pool = SqlitePoolOptions::new()
+			.max_connections(MAX_POOL_CONNECTIONS)
+			.connect(&url)
+			.await?;
+
+		// Create tables
+		sqlx::query(CREATE_TABLES_SQL).execute(&pool).await?;
+
+		Ok(Self { pool })
+	}
+
+	/// Open an in-memory cache (for testing).
+	pub async fn in_memory() -> Result<Self, CacheError> {
+		let pool = SqlitePoolOptions::new()
+			.max_connections(MEMORY_POOL_CONNECTIONS)
+			.connect(SQLITE_MEMORY_URL)
+			.await?;
+
+		sqlx::query(CREATE_TABLES_SQL).execute(&pool).await?;
+
+		Ok(Self { pool })
+	}
+
+	/// Get a cached storage value.
+	///
+	/// # Returns
+	/// * `Ok(Some(Some(value)))` - Cached with a value
+	/// * `Ok(Some(None))` - Cached as empty (storage key exists but has no value)
+	/// * `Ok(None)` - Not in cache (unknown)
+	pub async fn get(
+		&self,
+		block_hash: H256,
+		key: &[u8],
+	) -> Result<Option<Option<Vec<u8>>>, CacheError> {
+		// Retrieve the cached value and its empty flag for the given block and key.
+		// We need both `value` and `is_empty` to distinguish between:
+		// - Key not in cache (no row returned)
+		// - Key cached as empty (row exists, is_empty = true)
+		// - Key cached with value (row exists, is_empty = false)
+		let row =
+			sqlx::query("SELECT value, is_empty FROM storage WHERE block_hash = ? AND key = ?")
+				.bind(block_hash.as_bytes())
+				.bind(key)
+				.fetch_optional(&self.pool)
+				.await?;
+
+		Ok(row.map(|r| {
+			let is_empty: bool = r.get("is_empty");
+			if is_empty { None } else { Some(r.get("value")) }
+		}))
+	}
+
+	/// Cache a storage value.
+	///
+	/// # Arguments
+	/// * `block_hash` - The block hash this storage is from
+	/// * `key` - The storage key
+	/// * `value` - The storage value, or None if the key has no value (empty)
+	pub async fn set(
+		&self,
+		block_hash: H256,
+		key: &[u8],
+		value: Option<&[u8]>,
+	) -> Result<(), CacheError> {
+		// Insert or update the cached storage entry.
+		// Uses INSERT OR REPLACE (SQLite's UPSERT) to handle both new entries
+		// and updates to existing entries with the same (block_hash, key) primary key.
+		// The `is_empty` flag is set based on whether value is None, allowing us
+		// to cache the knowledge that a storage key has no value (vs not being cached).
+		sqlx::query(
+			"INSERT OR REPLACE INTO storage (block_hash, key, value, is_empty) VALUES (?, ?, ?, ?)",
+		)
+		.bind(block_hash.as_bytes())
+		.bind(key)
+		.bind(value)
+		.bind(value.is_none())
+		.execute(&self.pool)
+		.await?;
+
+		Ok(())
+	}
+
+	/// Get multiple cached storage values in a batch.
+	///
+	/// Returns results in the same order as the input keys.
+	pub async fn get_batch(
+		&self,
+		block_hash: H256,
+		keys: &[Vec<u8>],
```

Comment:
```suggestion
		keys: &[&[u8]],
```

We're using `key.as_slice()` inside the function, so we can directly pass slices as input and avoid forcing memory allocation when this is called, and is more flexible.

Also `get` uses `key: &[u8]`, so we should be consistent with that